src/ty.rs

```diff
@@ -77,29 +29,35 @@ pub trait WasmArg: WasmType {}
 /// Helper tait implemented by tuples to emulate "variadic generics".
 pub trait WasmArgs {
     #[doc(hidden)]
-    unsafe fn push_on_stack(self, stack: *mut [ffi::m3slot_t]);
+    unsafe fn push_on_stack(self, stack: *mut u64);
     #[doc(hidden)]
     // required for closure linking
-    unsafe fn pop_from_stack(stack: *mut [ffi::m3slot_t]) -> Self;
+    unsafe fn pop_from_stack(stack: *mut u64) -> Self;
     #[doc(hidden)]
-    fn validate_types(types: &[u8]) -> bool;
+    fn validate_types(types: &[ffi::M3ValueType::Type]) -> bool;
```

Comment:
```suggestion
    fn validate_types(types: impl Iterator<Item = ffi::M3ValueType::Type>) -> bool;
```
Doing this should allow us to get rid of the allocations annotated earlier